### ISSUE TITLE
IPv6

### ISSUE THREAD

**Issue by kseistrup** [HelloZeroNet/ZeroNet#148]
Please support IPv6.

**Comment by koalalorenzo** [comment]
+1

**Comment by HelloZeroNet** [comment]
The problem is the torrent trackers only support ipv4 addresses, so need to find an alternative solution to store ipv6 (and tor) addresses.

**Comment by bashrc** [comment]
I'm currently only using v4, but it looks like there is a BitTornado option to support v6

http://www.cs.helsinki.fi/u/sklvarjo/torrent.html

**Comment by up4** [comment]
I am working on something to make this easier: #520. Please check it out. Also, dumping BitTorrent trackers would be great in so many ways.

**Comment by iShift** [comment]
any progress? 
@HelloZeroNet this support ipv6 https://www.ipv6tracker.org

**Comment by HelloZeroNet** [comment]
It's not on short-term plans yet. (next 6 month)

**Comment by kseistrup** [comment]
[IAB Statement on IPv6](https://www.iab.org/documents/correspondence-reports-documents/2016-2/iab-statement-on-ipv6/):

> We recommend that existing standards be reviewed to ensure they will work with IPv6, and use IPv6 examples.

**Comment by eleitl** [comment]
Another vote for IPv6 support. Reason: using ZeroNet on top of cjdns/Hyperboria which is an IPv6-only transport layer.

**Comment by alugarius** [comment]
@HelloZeroNet  : ""It's not on short-term plans yet. (next 6 month)""
Narrator : ""7 months later...""
Me: "progress? "

**Comment by slothbag** [comment]
+1 for IPv6 support.. main reason is for CJDNS support.

**Comment by April93** [comment]
I'm also interested in IPV6 support, to let ZeroNet run on cjdns/hyperboria. I might take a crack at it a bit later, but I honestly have no idea how involved it'd be. I'd personally be up for no trackers and just having bootstrap nodes/peer exchange if the issue is that trackers aren't ipv6. IIRC, as it stands zeronet can work even if you remove the trackers from the code, because it stores peer info. So it shouldn't be too difficult to migrate.

**Comment by noxarivis** [comment]
1+ for above reasons

**Comment by grez911** [comment]
@HelloZeroNet Please, add some labels on this issue, it's important issue.

**Comment by PeterCxy** [comment]
> The problem is the torrent trackers only support ipv4  addresses, so need to find an alternative solution to store ipv6 (and  tor) addresses.

I'm not sure about the situation when this comment was made, but isn't it true that we now have the ZeroNet bootstrap protocol? Is it possible that IPv6 support could be added to this protocol like how it was done for Tor hosts?

**Comment by HelloZeroNet** [comment]
Yes, with ZeroNet bootstrap protocol It's no longer a problem.

**Comment by xuefer** [comment]
is it possible to add ipv6 support for web ui first?

**Comment by HelloZeroNet** [comment]
I think it should work if you start it with `--ui_ip youripv6address`

**Comment by April93** [comment]
Is this ready yet? How soon can we run it over cjdns? What all needs to be done?

**Comment by tangdou1** [comment]
Please support ipv6, most chinese college student have their free ipv6 but havt to pay for the ipv4. They share files with each other always using the free ipv6, not ipv4 because the ipv4 is paied by the track they used.

**Comment by tangdou1** [comment]
Here are the ipv6 tracker. Can it help?
udp://ipv6.tracker.harry.lu:80/announce
http://ipv6.tracker.harry.lu:80/announce
from
 https://github.com/ngosang/trackerslist/issues/72

**Comment by HelloZeroNet** [comment]
Yes, it can: were you able to find more?

**Comment by tangdou1** [comment]
I do not test these.

http://tracker.ipv6tracker.org:80/announce

udp://tracker.ipv6tracker.org:80/announce

from
https://www.makeuseof.com/answers/bittorrent-trackers-support-ipv6/

**Comment by tangdou1** [comment]
Ok. I found these now. Hope can help you.
http://0d.kebhana.mx:443/announce
udp://tracker.torrent.eu.org:451/announce
http://0d.kebhana.com.mx:443/announce
https://zer0day.000webhostapp.com:443/announce
udp://tracker.cypherpunks.ru:6969/announce
http://alpha.torrenttracker.nl:443/announce
https://1337.abcvg.info:443/announce
udp://free.open.tracker.0.starka.st:15480/announce
http://torrent.nwps.ws:80/announce
....
from https://newtrackon.com/

**Comment by tangdou1** [comment]
add
http://ipv6.torrent.ubuntu.com:6969/announce

**Comment by DaniellMesquita** [comment]
@HelloZeroNet @shortcutme
What is needed to ZeroNet support IPV6 and run through mesh networks?

**Comment by DaniellMesquita** [comment]
Reference: https://github.com/hyperboria/todo/issues/2

**Comment by April93** [comment]
@DaniellMesquita ZeroNet works on anything with IPv4 support. Just follow the bootstrapping instructions to set up a ZN-based tracker on your mesh. That's it. As for what mesh networks are ipv4.... I don't really know, sorry. I've been waiting on ipv6 support so I can use it with cjdns and hyperboria.

>With IPv6, ZeroNet can support CJDNS/Hyperboria mesh networks.

Indeed. That's why I'm eager for it.

**Comment by DaniellMesquita** [comment]
@April93 
WOW, is you running a CJDNS node?

**Comment by April93** [comment]
Any progress on this yet?

@DaniellMesquita I'm able to connect to cjdns, but I don't have anyone routing through me.

**Comment by DaniellMesquita** [comment]
@HelloZeroNet @shortcutme

**Comment by DaniellMesquita** [comment]
Reference: https://github.com/HelloZeroNet/ZeroNet/pull/1592

**Comment by tangdou1** [comment]
Good news for the official ipv6 support in rev.3802. It can be closed.

**Comment by DaniellMesquita** [comment]
@April93

**Comment by April93** [comment]
Anyone get cjdns ipv6 working? Still seems to be giving me problems.

**Comment by HelloZeroNet** [comment]
I have not tried but for now you have to start it with `--fileserver_ip_type ipv6` to bind on ipv6 interfaces instead of ipv4 ones. I still testing/monitoring it and will write about it on zeroblog.

**Comment by April93** [comment]
Yeah I noticed that it wasn't adding in my ipv6 trackers. I changed the setting in the configuration page and they popped up, but still don't connect. I feel like part of it is that I feel like I'm doing the bootstrapping thing wrong. I don't have regular ipv6 so I can't test the clearnet stuff. Which is what I assume it was added for.

Having cjdns working would be awesome.

**Comment by trenta3** [comment]
@tangdou1 From the new config page the setting on IPv6 says: "Accept other peers using IPv4 or IPv6 address. (default: IPv4)". What does it mean? If I enable IPv6 will I still be able to contact IPv4 peers or one must choose which type of addresses to use?

**Comment by April93** [comment]
I believe that just enables ipv6 support so it does both ipv4 and ipv6. As opposed to just ipv4 which is the default.

**Comment by HelloZeroNet** [comment]
You can connect to ipv6 and ipv4 peers in both mode. It changes what interface you will use for the incoming connections.

**Comment by HelloZeroNet** [comment]
If you have public ipv4 address, then set it to ipv4, if you have public ipv6 address, then set it to ipv6. If you don't have any, then doesn't really matter. 
I will see the possibility to add an "auto" mode that detect it on startup. It requires to start two file servers, which is not fully supported yet.

**Comment by kseistrup** [comment]
On Linux, binding to IPv6 (only) will allow the application to receive incoming connections on IPv6 _and_ IPv4.

```sh
$ sudo netstat -anp | grep ^tcp6 | tail -2
tcp6       0      0 192.168.1.50:15441      206.251.89.121:52549    ESTABLISHED 29569/zeronetd
tcp6       0      0 2001:570:28:4b5:::15441 2001:19f0:4400:75:34996 ESTABLISHED 29569/zeronetd
```

**Comment by HelloZeroNet** [comment]
Linux distributions with /proc/sys/net/ipv6/bindv6only disabled does binds on both ipv4/ipv6 automatically, Win10 doesn't, not sure about Mac.
Currently working on dual mode that will fire a proxy server to the other interface if necessary.

**Comment by HelloZeroNet** [comment]
I have added a dual stack mode by default to Rev3807: It binds to ipv6 interface if it's available and try to create a proxy server to the ipv4 interface (fails in most linux as ipv6 servers are also connectable with ipv4 there)
I also added modifications to support listening on both interfaces at the same time and a status indicator to  the port menu: 
![image](https://user-images.githubusercontent.com/10350359/51576792-cc7b6c80-1eb7-11e9-8a9a-ab0afca2b995.png)

**Comment by DaniellMesquita** [comment]
> Linux distributions with /proc/sys/net/ipv6/bindv6only disabled does binds on both ipv4/ipv6 automatically, Win10 doesn't, not sure about Mac.

@April93 could test that on Mac.
I'll try to update ZN and run my CJDNS node to test it.

**Comment by andrewheadricke** [comment]
I'm testing this out on CJDNS and am still having difficulties.  

Using both the `?zeronet_peer=` query string and the `trackers=` conf setting I don't think either of them support ip6 addresses.  You cant use the [ ] surrounding brackets, zeronet shows lots of errors.. but that's the only way to separate the address from the port.

**Comment by HelloZeroNet** [comment]
Right now addresses with brackets are not supported, the port is the part after the last ":" in the ipv6 address as you always have to define the port (as there is no default port for the peers)

**Comment by andrewheadricke** [comment]
```
IP Network: ['ipv4'] | Opened: {'ipv4': True, 'ipv6': None}
```

**Comment by andrewheadricke** [comment]
So `src/File/FileServer.py` has a check where it tries to connect to a hard coded IPv6 address to determine if its online.  Obviously a purely CJDNS node is going to fail that test.  I made the function just return true and now the node is online in IPv6 mode.  I can also see the nodes are successfully registering with my trackers.  But still both nodes are unable to find each other as peers.

**Comment by April93** [comment]
Just checked on my mac. Port is open for ipv4 as usual, says "unsupported" for ipv6. Switching the config page to ipv4 removes the ipv6 entries for the trackers, while ipv6 puts them there but fails to connect. The stats page lists `IP Network: ['ipv4'] | Opened: {'ipv4': True, 'ipv6': None}`.

I **don't** have native ipv6, but I **do** have cjdns which runs as a native ipv6 interface (pointing my browser to an ipv6 address navigates there properly, for instance).

I tried changing the hardcoded ipv6 test IP to a known cjdns hosted site that is up and put a log message for if ipv6 was indeed detected. Turns out that yes, the cjdns address worked fine for that hardcoded test (first line in isIpv6Supported function). However, even with this it still reported the port as closed, and I was unable to connect to myself as a tracker/bootstrapped node. So it's clear the connection method is proper (it detects that it can connect to the cjdns ip) but something is going wrong elsewhere. Perhaps I didn't set up my bootstrapper correctly? or a general incompatibility or messup somewhere else?

**Comment by HelloZeroNet** [comment]
I have changed the route test ip to a cjdns compatible one in Rev3808: https://github.com/HelloZeroNet/ZeroNet/commit/4d60c615126eb4726bed13dd3d4890864607a800

I have tried to install cjdns, but getting timeout error when I try to ping the ip.

Possible problems when using cjdns:
 - The port open check will fail as you won't be able to connect to the port checker site. To fix this we need a configuration option that allows the users to set the port open status manually.
 - No cjdns tracker in the default list.

**Comment by April93** [comment]
cjdns ips will _only_ work if you're running cjdns. Likewise native ipv6 will _only_ work if you have native ipv6. What cjdns does is route it's own ips and connect using inbuilt ipv6 support.

TL;DR:
If you only have cjdns you can only connect to cjdns ips.
If you only have native ipv6 you can only connect to native ipv6 ips.
If you have both then you can connect to both.

**Comment by HelloZeroNet** [comment]
> cjdns ips will only work if you're running cjdns. 

I did (I have no native ipv6 support from my isp), but still getting timeout error.

> As for the port check, yeah it needs some alternative way to check. As for tracker, shouldn't bootstrapping work?

It would, but right now the client only share the running tracker with other users if it has opened port. So if the port open check fails, then it won't share it by default.

**Comment by April93** [comment]
>I did (I have no native ipv6 support from my isp), but still getting timeout error.

You need to add a peer that's on hyperboria. I can give you info for that but I'm not at my pc right now. Just look up public cjdns peers and you should find some. The IP you used works for me. 

>It would, but right now the client only share the running tracker with other users if it has opened port. So if the port open check fails, then it won't share it by default.

Ah. Yeah that's an issue then.

**Comment by April93** [comment]
Here's a public cjdns peer:
```
"198.58.100.240:22237": {
                       "login": "default-login",
                       "password":"pqr5brz16vzzu6vhjuj7tv3n078kr5f",
                       "publicKey":"ubbtkp0txwjh44v8kkznvhjqqwr1hd2jzv5ms9zlkfk25svxvtg0.k",
                       "peerName":"trnsz"
                    },
```

**Comment by tangdou1** [comment]
> 
> just checked: actually right now the AnnounceShare only works with ipv4 address, I will fix it later today and also add an option to set the port open status manually.

It works. I find three ipv6 tracker in the  Shared trackers in my vps.

**Comment by DaniellMesquita** [comment]
If needing more CJDNS peers, here's mine:                                                                                                                                        

> "your.external.ip.goes.here:38064": {
>             "login": "default-login",
>             "password":"61px0khpzurswncqbkxcnfx9y7xfsyk",
>             "publicKey":"c7zwm2ykn6kuxpvdd18j476bhxvl6wn494jkxpx8cvkbc0x4bsq0.k",
>             "peerName":"daniell"
>         },

IP: `fc81:4668:5a19:0c29:9dfe:ed94:d745:85c9`

**Comment by DaniellMesquita** [comment]
To notice, I'm moving to Yggdrasil, so I'll report how it is working through ZeroNet.
@April93 how could I support both mesh networks?

**Comment by April93** [comment]
>Note: Why I need to provide an fixed IP address? That makes it centralized

You need an ip address to peer in cjdns just as you need an ip address to peer in zeronet. The method used to label your peer in cjdns doesn't need to be centralized. Ideally you'd have many peers, not just one.

>To notice, I'm moving to Yggdrasil, so I'll report how it is working through ZeroNet.
@April93 how could I support both mesh networks?

I'm not sure how you'd run both on your machine at once. Perhaps just run them? I don't know. For ZN once generic ipv6 support is in, then afaik yggdrasil should work fine as well.

**Comment by kseistrup** [comment]
Yggdrasil and cjdns runs fine side by side. I'm not sure how to make ZeroNet use them, though.

**Comment by kseistrup** [comment]
@DaniellMesquita,

> Why I need to provide an fixed IP address? That makes it centralized

Because otherwise peers won't know how to connect to your node and peer with it. If you didn't have any peers, your node would be just like a remote island that nobody knows exists. You need at least one peer.

**Comment by andrewheadricke** [comment]
I think there is some mix up over the "fixed ip address" thing.  The fixed ip address is referring to the connectivity check built into ZeroNet.  Personally I think it needs to be removed, or set to a configurable value because clearnet, cjdns, yggdrasil will all require separate values.

**Comment by andrewheadricke** [comment]
I have been trying to debug exactly why ZeroNet will not share IPv6 peers when running on cjdns/yggdrasil.  It appears to have something to do with the tracker exchange. I have one ZN node seeding a few zites, and here is the data it shares with the tracker..

```
"61663994dfd1003cae8ce419187841cd8a769d34":{"infoHash":"61663994dfd1003cae8ce419187841cd8a769d34","complete":0,"incomplete":1,"peers":{"_events":{},"_eventsCount":1,"cache":{"[fc12:a215:d106:2ca8:98eb:cae2:550f:6ce7]:15328":{"value":{"type":"http","complete":false,"peerId":"2d5554333533302d454253464f52366557483550","ip":"fc12:a215:d106:2ca8:98eb:cae2:550f:6ce7","port":15328},"modified":1548333888817,"next":null,"prev":null}},"tail":"[fc12:a215:d106:2ca8:98eb:cae2:550f:6ce7]:15328","head":"[fc12:a215:d106:2ca8:98eb:cae2:550f:6ce7]:15328","length":1,"max":1000,"maxAge":1200000}}
```

The only anomalies I can spot is the IP address contains the `[ ]` brackets which could cause a problem, and also the torrent is flagged as "incomplete" instead of complete. Which means the torrent is considered as leeching and not seeding.  Not sure if that could be the cause.

**Comment by HelloZeroNet** [comment]
Actually it doesn't matter if that IP is alive or not, because the function does not connects to it. (it's an UDP socket) It only used to test if the client's network stack has the ability to create a route towards to it.
We could add multiple ip addresses to it or as an alternative we can try to list the ip addresses on the network interface and use it to detect possible ipv6 support.

**Comment by HelloZeroNet** [comment]
I have modified the AnnounceShare to work with ipv6 addresses and support multiple ips in Rev3815 and added multiple public_ip support that is also configurable on /Config.
So if you add your cjdns ip to that field, then it should share your tracker on that ip with other users.

**Comment by April93** [comment]
ZN seems to be accepting the cjdns ip fine now (setting the config option). Trying to connect to my own machine as a tracker still isn't working. With the cjdns ip it stays at 0%, while with the default 0:0:0etc one it fluctuates but ultimately fails.

Not sure if I'm doing something wrong with the bootstrapping or i just need to connect to someone else.

**Comment by andrewheadricke** [comment]
Perhaps my external tracker is broken (i.e. doesn;t work well with ZeroNet).  

I now have a ZeroNet node on yggdrasil.  And I am able to connect to it from a ygg only node, and it appears to be working.

So to make this work I had to commend out the `check IPv6` function in `src/File/FileServer.py` as previously mentioned.  I then enabled the Bootstrapper plugin on my multi-homed node.  Then on the yggdrasil only node I added my tracker `zero://200:a215:d106:2ca8:98eb:cae2:550f:6ce7:15328` and it works.

Obviously it will only download the zites that I am seeding, which is ZeroSites, ZeroUp, ZeroName, GIF Time, ZeroHello, ZeroBlog and ZeroNet documentation.

The UI is not able to render the IPv6 tracker correctly, it cuts it off at the first semi colon.  But not a big issue.

**Comment by April93** [comment]
It looks like the ipv6 you provided isn't a cjdns ip, so cjdns peers can't connect to it. I don't know about yggdrasil or native ipv6 as I don't have either of those.

The zerohello page definitely cuts off the text of the ipv6 addresses, but it clearly still functions (being detected elsewhere fine). I don't think that impairs anything other than visuals.

**Comment by April93** [comment]
Great. Looks like it's connecting fine!

![screen shot 2019-01-25 at 8 08 58 pm](https://user-images.githubusercontent.com/27911816/51782449-52c4c800-20dd-11e9-95a7-a4b3aef6d423.png)

**Comment by DaniellMesquita** [comment]
@April93 I suggest you to try Yggdrasil, and you can run it alongside CJDNS.
Ygg is more scalable and more stable. Also, you can communicate with both networks.

I'll push the new commits and do feedback about if the @andrewheadricke's CJDNS tracker is working.

**Comment by HelloZeroNet** [comment]
I did other improvements to IPv6 in the past days:
 - It will consider as opened port (even if the port open check fails) if you have a non-local ip on your computer. (otherwise it would not add the client to tracker's peer list)
 - Consider any ipv6 ip as external except that starts with fe80 https://github.com/HelloZeroNet/ZeroNet/blob/master/src/util/helper.py#L252
 - Support connecting to IPv6 using Tor/Socks proxy. (this way you can connect to ipv6 even if your isp does not supports it)
 - You can see the detected external ips on the /Stats page. It should list your cjdns/Yggdrasil ip and it should detect your ipv6 as opened

**Comment by andrewheadricke** [comment]
For some reason my `/Stats` page wont load.. the console is showing some errors that I think are related.

```
[00:00:01] - UiWSGIHandler error: error: [Errno 32] Broken pipe in UiServer.py line 40 > pywsgi.py line 923 > pywsgi.py line 909 > pywsgi.py line 756 > pywsgi.py line 777 > pywsgi.py line 739 > pywsgi.py line 703 > _socket2.py line 354 > _socketcommon.py line 355 > _socketcommon.py line 284 > _socket2.py line 335
[00:00:54] - UiWSGIHandler error: TypeError: can't concat unicode to bytearray in UiServer.py line 40 > pywsgi.py line 923 > pywsgi.py line 909 > pywsgi.py line 752 > pywsgi.py line 736
```

**Comment by HelloZeroNet** [comment]
@andrewheadricke What OS are you using?
Also whats gevent version in your system?  It should be in log/debug.log file if you search for Gevent

**Comment by andrewheadricke** [comment]
Ubuntu Linux.  Gevent 1.3.6

**Comment by HelloZeroNet** [comment]
Thanks, I can confirm the error with gevent 1.3.6

**Comment by andrewheadricke** [comment]
I updated to gevent 1.4.0 and it seems the same problem exists.

**Comment by HelloZeroNet** [comment]
@andrewheadricke Fixed in Rev3831: https://github.com/HelloZeroNet/ZeroNet/commit/c2b879ccabc68cc4bceb81df373603547d4bfa77

**Comment by andrewheadricke** [comment]
Working well now, thanks.

I feel there are still a few small issues with IPv6 support, but technically its working and I have a 24/7 tracker online on Yggdrasil at `zero://203:5871:3e33:413c:a824:cb7b:2604:92ab:24589` and it appears to be running fine.

I vote this issue can be closed :)

**Comment by April93** [comment]
Yeah it's working great. This can be closed now.
